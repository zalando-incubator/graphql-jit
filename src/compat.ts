import {
  GraphQLSchema,
  GraphQLError,
  versionInfo,
  FieldNode,
  GraphQLField
} from "graphql";
import { GraphQLObjectType } from "graphql/type/definition";
import { Maybe } from "graphql/jsutils/Maybe";

import { ASTNode, OperationDefinitionNode } from "graphql/language/ast";
import * as errorUtilities from "graphql/error";
import * as utilities from "graphql/utilities";
import { GraphQLFormattedError } from "graphql/error";
import { CompilationContext } from "./execution";
import * as execute from "graphql/execution/execute";

/**
 * A helper file to support backward compatibility for different versions of graphql-js.
 */

/**
 * v15 does not have schema.getRootType
 * v16 has both
 * v17 will not have getOperationRootType
 *
 * To support all these 3 versions of graphql-js, at least for migration, this helper
 * would be useful.
 *
 * This can be removed once we drop support for v15.
 *
 * GraphQL v17 would remove getOperationRootType.
 */
export function getOperationRootType(
  schema: GraphQLSchema,
  operation: OperationDefinitionNode
): GraphQLObjectType {
  if (versionInfo.major < 16) {
    return (utilities as any).getOperationRootType(schema, operation);
  }

  const type = (schema as any).getRootType(operation.operation);

  if (!type) {
    throw new Error(`No root type for operation ${operation.operation}`);
  }

  return type;
}

/**
 * v16 and lower versions don't have .toJSON method on GraphQLError
 * v17 does have .toJSON and doesn't have "formatError" export anymore
 */
export function formatError(error: GraphQLError): GraphQLFormattedError {
  if (versionInfo.major < 16) {
    return (errorUtilities as any).formatError(error);
  }

  return (error as any).toJSON();
}

/**
 * v17 dropped support for positional arguments in GraphQLError constructor
 * https://github.com/graphql/graphql-js/pull/3577
 */
export function getGraphQLErrorOptions(
  nodes: Maybe<ReadonlyArray<ASTNode> | ASTNode>
): ConstructorParameters<typeof GraphQLError>[1] {
  if (versionInfo.major < 16) {
    return nodes as any;
  }

  return { nodes } as any;
}

/**
 * Resolves the field on the given source object. In particular, this
 * figures out the value that the field returns by calling its resolve function,
 * then calls completeValue to complete promises, serialize scalars, or execute
 * the sub-selection-set for objects.
 *
 * v15 has getFieldDef that accepts field name
 * v16 has getFieldDef that accepts field node
 * v17 drops getFieldDef support and adds getField method
 */
export function resolveFieldDef(
  compilationContext: CompilationContext,
  parentType: GraphQLObjectType,
  fieldNodes: FieldNode[]
): Maybe<GraphQLField<any, any>> {
  const fieldNode = fieldNodes[0];

  if (versionInfo.major < 16) {
    const fieldName = fieldNode.name.value;
    return (execute as any).getFieldDef(
      compilationContext.schema,
      parentType,
      fieldName as any
    );
  }

  if (versionInfo.major < 17) {
    return (execute as any).getFieldDef(
      compilationContext.schema,
      parentType,
      fieldNode as any
    );
  }

  return (compilationContext.schema as any).getField(
    parentType,
    fieldNode.name.value
  );
}
