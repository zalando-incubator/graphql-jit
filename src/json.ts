/**
 * Mapping between GQL primitive types and JSON Schema property types
 *
 * @type       {<type>}
 */
import {
  FieldNode,
  getOperationRootType,
  GraphQLType,
  isAbstractType,
  isEnumType,
  isListType,
  isNonNullType,
  isObjectType,
  isScalarType
} from "graphql";
import { JSONSchema6, JSONSchema6TypeName } from "json-schema";
import { collectFields, collectSubfields, resolveFieldDef } from "./ast";
import { CompilationContext } from "./execution";

const PRIMITIVES: { [key: string]: JSONSchema6TypeName } = {
  Int: "integer",
  Float: "number",
  String: "string",
  Boolean: "boolean",
  ID: "string"
};

/**
 * GQL -> JSON Schema transform
 *
 * @param compilationContext
 * @return     {object}  A plain JavaScript object which conforms to JSON Schema
 */
export function queryToJSONSchema(
  compilationContext: CompilationContext
): JSONSchema6 {
  const type = getOperationRootType(
    compilationContext.schema,
    compilationContext.operation
  );
  const fields = collectFields(
    compilationContext,
    type,
    compilationContext.operation.selectionSet,
    Object.create(null),
    Object.create(null)
  );
  const fieldProperties = Object.create(null);
  for (const responseName of Object.keys(fields)) {
    const fieldType = resolveFieldDef(
      compilationContext,
      type,
      fields[responseName]
    );
    if (!fieldType) {
      // if field does not exist, it should be ignored for compatibility concerns.
      // Usually, validation would stop it before getting here but this could be an old query
      continue;
    }
    fieldProperties[responseName] = transformNode(
      compilationContext,
      fields[responseName],
      fieldType.type
    );
  }
  return {
    type: "object",
    properties: {
      data: {
        type: ["object", "null"],
        properties: fieldProperties
      },
      errors: {
        type: "array",
        items: {
          type: "object",
          additionalProperties: true,
          properties: {
            message: {
              type: "string"
            },
            path: {
              type: "array",
              items: {
                type: ["string", "number"]
              }
            },
            locations: {
              type: "array",
              items: {
                type: "object",
                properties: {
                  line: {
                    type: "number"
                  },
                  column: {
                    type: "number"
                  }
                }
              }
            }
          }
        }
      }
    }
  };
}

function transformNode(
  compilationContext: CompilationContext,
  fieldNodes: FieldNode[],
  type: GraphQLType
): JSONSchema6 {
  if (isObjectType(type)) {
    const subfields = collectSubfields(compilationContext, type, fieldNodes);
    const properties = Object.create(null);
    for (const responseName of Object.keys(subfields)) {
      const fieldType = resolveFieldDef(
        compilationContext,
        type,
        subfields[responseName]
      );
      if (!fieldType) {
        // if field does not exist, it should be ignored for compatibility concerns.
        // Usually, validation would stop it before getting here but this could be an old query
        continue;
      }
      properties[responseName] = transformNode(
        compilationContext,
        subfields[responseName],
        fieldType.type
      );
    }
    return {
      type: ["object", "null"],
      properties
    };
  }
  if (isListType(type)) {
    return {
      type: ["array", "null"],
      items: transformNode(compilationContext, fieldNodes, type.ofType)
    };
  }
  if (isNonNullType(type)) {
    const nullable = transformNode(compilationContext, fieldNodes, type.ofType);
    if (nullable.type && Array.isArray(nullable.type)) {
      const nonNullable = nullable.type.filter((x) => x !== "null");
      return {
        ...nullable,
        type: nonNullable.length === 1 ? nonNullable[0] : nonNullable
      };
    }
    return {};
  }
  if (isEnumType(type)) {
    return {
      type: ["string", "null"]
    };
  }
  if (isScalarType(type)) {
    const jsonSchemaType = PRIMITIVES[type.name];
    if (!jsonSchemaType) {
      return {};
    }
    return {
      type: [jsonSchemaType, "null"]
    };
  }
  if (isAbstractType(type)) {
    return compilationContext.schema.getPossibleTypes(type).reduce(
      (res, t) => {
        const jsonSchema = transformNode(compilationContext, fieldNodes, t);
        res.properties = { ...res.properties, ...jsonSchema.properties };
        return res;
      },
      {
        type: ["object", "null"],
        properties: {}
      } as JSONSchema6
    );
  }
  throw new Error(`Got unhandled type: ${type.name}`);
}
