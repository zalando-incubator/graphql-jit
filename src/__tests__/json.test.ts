import fastJson from "fast-json-stringify";
import {
  GraphQLBoolean,
  GraphQLError,
  GraphQLID,
  GraphQLList,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLString,
  parse,
  GraphQLInt,
  versionInfo
} from "graphql";
import { buildExecutionContext } from "graphql/execution/execute";
import { compileQuery } from "../index";
import { queryToJSONSchema } from "../json";
import { makeExecutableSchema } from "@graphql-tools/schema";
import { formatError } from "../compat";

describe("json schema creator", () => {
  const BlogAuthor = new GraphQLObjectType({
    name: "Author",
    fields: () => ({
      id: { type: new GraphQLNonNull(GraphQLID) },
      name: { type: GraphQLString },
      pic: {
        type: new GraphQLObjectType({
          name: "Pic",
          fields: {
            width: { type: GraphQLInt },
            height: { type: GraphQLInt },
            url: { type: GraphQLString }
          }
        })
      },
      recentArticle: {
        type: BlogArticle
      }
    })
  });

  const BlogArticle: GraphQLObjectType = new GraphQLObjectType({
    name: "Article",
    fields: {
      id: { type: new GraphQLNonNull(GraphQLID) },
      isPublished: { type: GraphQLBoolean },
      author: { type: new GraphQLNonNull(BlogAuthor) },
      title: { type: GraphQLString },
      body: { type: GraphQLString }
    }
  });

  const BlogQuery = new GraphQLObjectType({
    name: "Query",
    fields: {
      article: {
        type: BlogArticle,
        args: { id: { type: GraphQLID } },
        resolve: (_, { id }) => article(id)
      },
      feed: {
        type: new GraphQLList(BlogArticle),
        resolve: () => [
          article(1),
          article(2),
          article(3),
          article(4),
          article(5),
          article(6),
          article(7),
          article(8),
          article(9),
          article(10)
        ]
      }
    }
  });

  function article(id: number): any {
    return {
      id,
      isPublished: true,
      author: {
        id: 123,
        name: "John Smith"
      },
      title: "My Article " + id,
      body: "This is a post",
      hidden: "This data is not exposed in the schema"
    };
  }

  const blogSchema = new GraphQLSchema({
    query: BlogQuery
  });

  const document = parse(/* GraphQL */ `
    {
      feed {
        id
        title
      }
      article(id: "1") {
        ...articleFields
        author {
          id
          name
          pic(width: 640, height: 480) {
            url
            width
            height
          }
          recentArticle {
            ...articleFields
            keywords
          }
        }
      }
    }

    fragment articleFields on Article {
      id
      isPublished
      title
      body
      hidden
      notdefined
    }
  `);

  const context: any =
    versionInfo.major > 15
      ? (buildExecutionContext as any)({
          schema: blogSchema,
          document
        })
      : (buildExecutionContext as any)(blogSchema, document);
  const jsonSchema = queryToJSONSchema(context);
  test("json schema creation", () => {
    expect(jsonSchema).toMatchSnapshot();
  });

  describe("fast json stringify", () => {
    test("valid json schema", () => {
      expect(typeof fastJson(jsonSchema) === "function").toBeTruthy();
    });
    test("valid response serialization", async () => {
      const prepared: any = compileQuery(blogSchema, document, "", {
        customJSONSerializer: true
      });
      const response = await prepared.query(undefined, undefined, {});
      expect(prepared.stringify).not.toBe(JSON.stringify);
      expect(prepared.stringify(response)).toEqual(JSON.stringify(response));
    });
    test("valid response serialization 2", async () => {
      const prepared: any = compileQuery(blogSchema, document, "", {
        customJSONSerializer: false
      });
      expect(prepared.stringify).toBe(JSON.stringify);
    });
    test("error response serialization", async () => {
      const stringify = fastJson(jsonSchema);
      const response = {
        errors: [formatError(new GraphQLError("test"))]
      };
      expect(stringify(response)).toEqual(JSON.stringify(response));
    });
  });
});

describe("JSON schema creation with abstract types", () => {
  let inf: any;
  const schema = makeExecutableSchema({
    typeDefs: `
        type Query {
          foo: Foo
          u: U
        }
        union U = Foo | Bar

        type Foo {
          a: String
          b: Int
          c: Boolean!
          d: Bar
        }
        type Bar {
          e: String!
          f: Boolean!
        }
      `,
    resolvers: {
      Query: {
        foo(_: any, _1: any, _2: any, info: any) {
          inf = info;
        },
        u(_: any, _1: any, _2: any, info: any) {
          inf = info;
        }
      }
    }
  });
  const document = parse(/* GraphQL */ `
    {
      u
    }
  `);
  const context: any =
    versionInfo.major > 15
      ? (buildExecutionContext as any)({
          schema,
          document
        })
      : (buildExecutionContext as any)(schema, document);

  const jsonSchema = queryToJSONSchema(context);
  test("json schema creation", () => {
    expect(jsonSchema).toMatchSnapshot();
  });
});
