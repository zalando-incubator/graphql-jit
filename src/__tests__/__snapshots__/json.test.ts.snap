// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSON schema creation with abstract types json schema creation 1`] = `
Object {
  "properties": Object {
    "data": Object {
      "properties": Object {
        "u": Object {
          "properties": Object {},
          "type": Array [
            "object",
            "null",
          ],
        },
      },
      "type": Array [
        "object",
        "null",
      ],
    },
    "errors": Object {
      "items": Object {
        "additionalProperties": true,
        "properties": Object {
          "locations": Object {
            "items": Object {
              "properties": Object {
                "column": Object {
                  "type": "number",
                },
                "line": Object {
                  "type": "number",
                },
              },
              "type": "object",
            },
            "type": "array",
          },
          "message": Object {
            "type": "string",
          },
          "path": Object {
            "items": Object {
              "type": Array [
                "string",
                "number",
              ],
            },
            "type": "array",
          },
        },
        "type": "object",
      },
      "type": "array",
    },
  },
  "type": "object",
}
`;

exports[`json schema creator json schema creation 1`] = `
Object {
  "properties": Object {
    "data": Object {
      "properties": Object {
        "article": Object {
          "properties": Object {
            "author": Object {
              "properties": Object {
                "id": Object {
                  "type": "string",
                },
                "name": Object {
                  "type": Array [
                    "string",
                    "null",
                  ],
                },
                "pic": Object {
                  "properties": Object {
                    "height": Object {
                      "type": Array [
                        "integer",
                        "null",
                      ],
                    },
                    "url": Object {
                      "type": Array [
                        "string",
                        "null",
                      ],
                    },
                    "width": Object {
                      "type": Array [
                        "integer",
                        "null",
                      ],
                    },
                  },
                  "type": Array [
                    "object",
                    "null",
                  ],
                },
                "recentArticle": Object {
                  "properties": Object {
                    "body": Object {
                      "type": Array [
                        "string",
                        "null",
                      ],
                    },
                    "id": Object {
                      "type": "string",
                    },
                    "isPublished": Object {
                      "type": Array [
                        "boolean",
                        "null",
                      ],
                    },
                    "title": Object {
                      "type": Array [
                        "string",
                        "null",
                      ],
                    },
                  },
                  "type": Array [
                    "object",
                    "null",
                  ],
                },
              },
              "type": "object",
            },
            "body": Object {
              "type": Array [
                "string",
                "null",
              ],
            },
            "id": Object {
              "type": "string",
            },
            "isPublished": Object {
              "type": Array [
                "boolean",
                "null",
              ],
            },
            "title": Object {
              "type": Array [
                "string",
                "null",
              ],
            },
          },
          "type": Array [
            "object",
            "null",
          ],
        },
        "feed": Object {
          "items": Object {
            "properties": Object {
              "id": Object {
                "type": "string",
              },
              "title": Object {
                "type": Array [
                  "string",
                  "null",
                ],
              },
            },
            "type": Array [
              "object",
              "null",
            ],
          },
          "type": Array [
            "array",
            "null",
          ],
        },
      },
      "type": Array [
        "object",
        "null",
      ],
    },
    "errors": Object {
      "items": Object {
        "additionalProperties": true,
        "properties": Object {
          "locations": Object {
            "items": Object {
              "properties": Object {
                "column": Object {
                  "type": "number",
                },
                "line": Object {
                  "type": "number",
                },
              },
              "type": "object",
            },
            "type": "array",
          },
          "message": Object {
            "type": "string",
          },
          "path": Object {
            "items": Object {
              "type": Array [
                "string",
                "number",
              ],
            },
            "type": "array",
          },
        },
        "type": "object",
      },
      "type": "array",
    },
  },
  "type": "object",
}
`;
